// vite-plugin-spa/src/spa-view.ts
import { defineComponent, h, nextTick, onMounted, Ref, ref, watch } from 'vue';

interface ApaApp {
  id: string;
  show: boolean;
}

export default defineComponent({
  name: 'Vite-Spa',
  props: {
    app: String
  },
  setup(props) {
    const apps: Ref<ApaApp[]> = ref([]);
    const renderApp = (appName: string) => {
      apps.value.map((app) => (app.show = false));
      console.log(apps.value);
      const showedApp = apps.value.find((app) => app.id === appName);
      if (showedApp) {
        console.log(showedApp);
        showedApp.show = true;
        return;
      }

      apps.value.push({ id: appName, show: true });
      nextTick(() => {
        fetch(`/${appName}/manifest.json`)
          .then((response) => response.json())
          .then((data) => {
            let src = data['index.html']?.file;
            if (!/^\/\S/.test(src)) {
              src = `/${src}`;
            }
            const script = document.createElement('script');
            script.type = 'module';
            script.src = `/${appName}${src}`;
            const body = document.querySelector(appName);
            body?.appendChild(script);
          });
      });
    };

    onMounted(() => {});
    watch(
      () => props.app,
      (appName) => {
        console.log(appName);
        appName && renderApp(appName);
      },
      { immediate: true }
    );
    return { apps };
  },
  render() {
    return this.apps.map((app) => {
      const attrs: any = JSON.parse(JSON.stringify(this.$attrs));
      attrs.id = app.id;
      if (!attrs.style) {
        attrs.style = { display: '' };
      }
      attrs.style.display = app.show ? '' : 'none';
      console.log(attrs);
      return h(app.id, attrs);
    });
  }
});
