// vite-plugin-spa/src/vite-plugin-spa.ts
import { Plugin } from 'vite';

interface PluginOptions {
  scriptSrc?: string;
}
const uuid = () =>{
    let s:any = [];
    const hexDigits = '0123456789abcdef';
    for (var i = 0; i < 32; i++) {
      s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
    }
    s[14] = '4';
    s[19] = hexDigits.substr((s[19] & 0x3) | 0x8, 1);
    s[8] = s[13] = s[18] = s[23];
    return s.join('');
  }
export default function vitePluginSpa(options?: PluginOptions): Plugin {
  return {
    name: 'vite-plugin-spa',

    configureServer(server) {
      server.middlewares.use((req, res, next) => {
        if (req.originalUrl.includes('manifest.json')) {
          const scriptSrc = options?.scriptSrc;
          const src = {
            'index.html': {
              file: scriptSrc || '/src/main.ts',
              src: 'index.html'
            }
          };
          res.end(JSON.stringify(src));
        } else {
          next();
        }
      });
    },

    transform: (src, id) => {
      const overlayId = 'vite-error-overlay';
      if (src.includes(overlayId)) {
        return {
          code: src.replace(overlayId, overlayId + uuid()),
          map: { mappings: '' }
        };
      }
    }
  };
}
